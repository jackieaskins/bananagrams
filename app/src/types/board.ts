import { Tile } from "./tile";

export type BoardLocation = {
  x: number;
  y: number;
};

export enum Direction {
  ACROSS = "ACROSS",
  DOWN = "DOWN",
}

export enum ValidationStatus {
  NOT_VALIDATED = "NOT_VALIDATED",
  VALID = "VALID",
  INVALID = "INVALID",
}

export type WordInfo = {
  start: BoardLocation;
  validation: ValidationStatus;
};

export type AbstractBoardSquare<T> = {
  tile: T;
  wordInfo: Record<Direction, WordInfo>;
};

export type AbstractBoard<T> = Record<string, AbstractBoardSquare<T>>;

export type BoardSquare = AbstractBoardSquare<Tile>;

export type Board = AbstractBoard<Tile>;
