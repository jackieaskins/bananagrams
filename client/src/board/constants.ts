import { TILE_SIZE } from '../tile/constants';

const SCREEN_MULTIPLIER = 1.5;
const MIN_TILES = 25;

const getNumTiles = (len: number): number =>
  Math.max(Math.ceil((len * SCREEN_MULTIPLIER) / TILE_SIZE), MIN_TILES);
export const TILES_PER_ROW = getNumTiles(window.screen.width);
export const TILES_PER_COL = getNumTiles(window.screen.height);

export const BOARD_WIDTH = TILES_PER_ROW * TILE_SIZE + 2;
export const BOARD_HEIGHT = TILES_PER_COL * TILE_SIZE + 2;

export const BORDER_STYLE = '1px solid rgba(173, 216, 230, 0.5)';
