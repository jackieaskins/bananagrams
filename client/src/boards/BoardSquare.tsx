/* eslint-disable @typescript-eslint/ban-ts-comment */
import { Box } from '@material-ui/core';
import { useDrop } from 'react-dnd';

import { useGame } from '../games/GameContext';
import { useStyles } from '../styles';
import Tile from '../tiles/Tile';
import { TileItem } from '../tiles/types';
import {
  BoardSquare as BoardSquareType,
  Direction,
  WordInfo,
  ValidationStatus,
} from './types';

type BoardSquareProps = {
  boardSquare: BoardSquareType | null;
  row: number;
  col: number;
};

type CheckValidation = (wordInfo: WordInfo) => boolean;

const getColor = (
  wordInfo: Record<Direction, WordInfo>
): string | undefined => {
  const isValid: CheckValidation = ({ validation }) =>
    validation === ValidationStatus.VALID;
  const isValidated: CheckValidation = ({ validation }) =>
    validation !== ValidationStatus.NOT_VALIDATED;

  const validations = Object.values(wordInfo).filter(isValidated);

  if (validations.length === 0) return 'black';
  if (validations.every(isValid)) return 'green';
  return 'red';
};

const BoardSquare = ({
  boardSquare,
  row,
  col,
}: BoardSquareProps): JSX.Element => {
  const classes = useStyles();
  const { handleMoveTileFromHandToBoard, handleMoveTileOnBoard } = useGame();
  const { tile, wordInfo } = boardSquare ?? {};

  const [{ canDrop, isOver }, dropRef] = useDrop({
    accept: 'TILE',
    canDrop: (_, monitor) => monitor.isOver() && !tile,
    drop: ({ id, boardPosition }: TileItem) => {
      if (!!boardPosition) {
        handleMoveTileOnBoard(boardPosition, { row, col });
      } else {
        handleMoveTileFromHandToBoard(id, { row, col });
      }
    },
    collect: (monitor) => ({
      isOver: monitor.isOver({ shallow: true }),
      canDrop: monitor.canDrop(),
    }),
  });

  return (
    <Box
      className={isOver && canDrop ? classes.validDrop : ''}
      height="25px"
      width="25px"
      // @ts-ignore
      ref={dropRef}
    >
      {tile && wordInfo ? (
        <Tile
          tile={tile}
          color={getColor(wordInfo)}
          boardPosition={{ row, col }}
        />
      ) : (
        ''
      )}
    </Box>
  );
};

export default BoardSquare;
