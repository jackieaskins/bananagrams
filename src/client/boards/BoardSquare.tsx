import { Box } from "@chakra-ui/react";
import { useDrop } from "react-dnd";
import { useEnableTileSwap } from "../LocalStorageContext";
import { useGame } from "../games/GameContext";
import Tile from "../tiles/Tile";
import { TileItem } from "../tiles/types";
import {
  BoardSquare as BoardSquareType,
  Direction,
  ValidationStatus,
  WordInfo,
} from "./types";

type BoardSquareProps = {
  boardSquare: BoardSquareType | null;
  x: number;
  y: number;
};

type CheckValidation = (wordInfo: WordInfo) => boolean;

function getColor(wordInfo: Record<Direction, WordInfo>): string | undefined {
  const isValid: CheckValidation = ({ validation }) =>
    validation === ValidationStatus.VALID;
  const isValidated: CheckValidation = ({ validation }) =>
    validation !== ValidationStatus.NOT_VALIDATED;

  const validations = Object.values(wordInfo).filter(isValidated);

  if (validations.length === 0) return "black";
  if (validations.every(isValid)) return "green.500";
  return "red.600";
}

export default function BoardSquare({
  boardSquare,
  x,
  y,
}: BoardSquareProps): JSX.Element {
  const [enableTileSwap] = useEnableTileSwap();
  const { handleMoveTileFromHandToBoard, handleMoveTileOnBoard } = useGame();
  const { tile, wordInfo } = boardSquare ?? {};

  const [{ canDrop, isOver }, dropRef] = useDrop({
    accept: "TILE",
    canDrop: (_, monitor) => monitor.isOver() && (enableTileSwap || !tile),
    drop: ({ id, boardLocation }: TileItem) => {
      if (boardLocation) {
        handleMoveTileOnBoard(boardLocation, { x, y });
      } else {
        handleMoveTileFromHandToBoard(id, { x, y });
      }
    },
    collect: (monitor) => ({
      isOver: monitor.isOver({ shallow: true }),
      canDrop: monitor.canDrop(),
    }),
  });

  return (
    <Box
      backgroundColor={isOver && canDrop ? "green.400" : undefined}
      color={isOver && canDrop ? "green.400" : undefined}
      opacity={isOver && canDrop ? 0.5 : 1}
      height="25px"
      width="25px"
      ref={dropRef}
    >
      {tile && wordInfo ? (
        <Tile tile={tile} color={getColor(wordInfo)} boardLocation={{ x, y }} />
      ) : (
        ""
      )}
    </Box>
  );
}
