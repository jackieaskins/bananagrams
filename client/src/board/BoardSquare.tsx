import { useRecoilCallback } from 'recoil';

import { BoardPosition } from '../boards/types';
import { useCurrentBoardSquare } from '../game/stateHooks';
import { moveTile } from '../socket';
import Tile from '../tile/Tile';
import { TILE_SIZE } from '../tile/constants';
import { selectedTileState } from '../tile/selectedTileState';
import { BORDER_STYLE } from './constants';

const BoardSquare = ({ row, col }: BoardPosition): JSX.Element => {
  const square = useCurrentBoardSquare({ row, col });

  const handleClick = useRecoilCallback(
    ({ set, snapshot }) => async () => {
      const selectedTile = await snapshot.getPromise(selectedTileState);

      if (selectedTile) {
        moveTile({
          tileId: selectedTile.tile.id,
          fromPosition: selectedTile.boardPosition,
          toPosition: { row, col },
        });

        set(selectedTileState, null);
      }
    },
    [row, col]
  );

  const selectTile = useRecoilCallback(
    ({ set, snapshot }) => async ({ tile }) => {
      const selectedTile = await snapshot.getPromise(selectedTileState);

      if (!selectedTile) {
        set(selectedTileState, { tile, boardPosition: { row, col } });
      }
    },
    [col, row]
  );

  return (
    <div
      css={{
        minWidth: `${TILE_SIZE}px`,
        width: `${TILE_SIZE}px`,
        height: `${TILE_SIZE}px`,
        border: BORDER_STYLE,
        display: 'flex',
        alignItems: 'center',
        justifyContent: 'center',
      }}
      onClick={handleClick}
    >
      {square ? (
        <Tile
          id={square.tile.id}
          letter={square.tile.letter}
          onClick={selectTile}
          validationStatus={square.validationStatus}
        />
      ) : null}
    </div>
  );
};

export default BoardSquare;
