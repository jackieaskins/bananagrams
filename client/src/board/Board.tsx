import { useMemo } from 'react';
import Draggable from 'react-draggable';

import BoardSquare from './BoardSquare';
import {
  BOARD_WIDTH,
  BOARD_HEIGHT,
  BORDER_STYLE,
  TILES_PER_COL,
  TILES_PER_ROW,
} from './constants';

const Board = (): JSX.Element => {
  const boardSquares = useMemo(
    () =>
      [...Array(TILES_PER_COL)].map((_, y) => (
        <div key={y} css={{ display: 'flex' }}>
          {[...Array(TILES_PER_ROW)].map((_, x) => (
            <BoardSquare key={x} />
          ))}
        </div>
      )),
    []
  );

  return (
    <Draggable bounds="parent" cancel=".no-drag">
      <div
        css={{
          cursor: 'move',
          display: 'flex',
          flexDirection: 'column',
          width: `${BOARD_WIDTH}px`,
          height: `${BOARD_HEIGHT}px`,
          border: BORDER_STYLE,
        }}
      >
        {boardSquares}
      </div>
    </Draggable>
  );
};

export default Board;
