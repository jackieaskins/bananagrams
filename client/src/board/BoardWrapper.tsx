import { useWindowSize } from '../util/window';
import Board from './Board';
import { BOARD_HEIGHT, BOARD_WIDTH } from './constants';

const BoardView = (): JSX.Element => {
  const { width: windowWidth, height: windowHeight } = useWindowSize();

  return (
    <div
      css={{
        position: 'fixed',
        left:
          BOARD_WIDTH > windowWidth
            ? windowWidth - BOARD_WIDTH - 5
            : (windowWidth - BOARD_WIDTH) / 2,
        right:
          BOARD_WIDTH > windowWidth
            ? windowWidth - BOARD_WIDTH - 5
            : (windowWidth - BOARD_WIDTH) / 2,
        top:
          BOARD_HEIGHT > windowHeight
            ? windowHeight - BOARD_HEIGHT - 5
            : (windowHeight - BOARD_HEIGHT) / 2,
        bottom:
          BOARD_HEIGHT > windowHeight
            ? windowHeight - BOARD_HEIGHT - 5
            : (windowHeight - BOARD_HEIGHT) / 2,
        display: 'flex',
        alignItems: 'center',
        justifyContent: 'center',
      }}
    >
      <Board />
    </div>
  );
};

export default BoardView;
