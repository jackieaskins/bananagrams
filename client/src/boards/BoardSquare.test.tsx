/* eslint-disable testing-library/no-render-in-lifecycle */
import { shallow } from "enzyme";
import { useDrop } from "react-dnd";
import { boardSquareFixture, wordInfoFixture } from "../fixtures/board";
import BoardSquare from "./BoardSquare";
import {
  BoardSquare as BoardSquareType,
  Direction,
  ValidationStatus,
  WordInfo,
  BoardLocation,
} from "./types";

jest.mock("react-dnd", () => ({
  useDrop: jest.fn(),
}));

const mockHandleMoveTileFromHandToBoard = jest.fn();
const mockHandleMoveTileOnBoard = jest.fn();
jest.mock("../games/GameContext", () => ({
  useGame: () => ({
    handleMoveTileFromHandToBoard: mockHandleMoveTileFromHandToBoard,
    handleMoveTileOnBoard: mockHandleMoveTileOnBoard,
  }),
}));

describe("<BoardSquare />", () => {
  const getBoardSquare = (
    across: WordInfo = wordInfoFixture(),
    down: WordInfo = wordInfoFixture(),
  ): BoardSquareType =>
    boardSquareFixture({
      wordInfo: {
        [Direction.ACROSS]: across,
        [Direction.DOWN]: down,
      },
    });

  const renderComponent = (propOverrides = {}) =>
    shallow(<BoardSquare boardSquare={null} x={0} y={0} {...propOverrides} />);

  beforeEach(() => {
    useDrop.mockReturnValue([{ canDrop: true, isOver: false }, jest.fn()]);
  });

  test("renders properly when is over and can drop", () => {
    useDrop.mockReturnValue([{ canDrop: true, isOver: true }, jest.fn()]);

    expect(renderComponent()).toMatchSnapshot();
  });

  test("renders properly with null boardSquare", () => {
    expect(renderComponent()).toMatchSnapshot();
  });

  test("renders black tile when not validated in any direction", () => {
    const boardSquare = getBoardSquare(
      wordInfoFixture({ validation: ValidationStatus.NOT_VALIDATED }),
      wordInfoFixture({ validation: ValidationStatus.NOT_VALIDATED }),
    );

    expect(renderComponent({ boardSquare })).toMatchSnapshot();
  });

  test("renders green tile when valid in every direction", () => {
    const boardSquare = getBoardSquare(
      wordInfoFixture({ validation: ValidationStatus.VALID }),
      wordInfoFixture({ validation: ValidationStatus.VALID }),
    );

    expect(renderComponent({ boardSquare })).toMatchSnapshot();
  });

  test("renders red tile when invalid in any direction", () => {
    const boardSquare = getBoardSquare(
      wordInfoFixture({ validation: ValidationStatus.VALID }),
      wordInfoFixture({ validation: ValidationStatus.INVALID }),
    );

    expect(renderComponent({ boardSquare })).toMatchSnapshot();
  });

  describe("useDrop", () => {
    test("is called", () => {
      renderComponent();

      expect(useDrop).toHaveBeenCalledWith({
        accept: "TILE",
        canDrop: expect.any(Function),
        drop: expect.any(Function),
        collect: expect.any(Function),
      });
    });

    describe("canDrop", () => {
      const callCanDrop = (monitor) =>
        useDrop.mock.calls[0][0].canDrop({}, monitor);

      test("returns true if monitor is over and no tile exists", () => {
        renderComponent();

        expect(
          callCanDrop({
            isOver: () => true,
          }),
        ).toBe(true);
      });

      test("returns false if monitor is not over but no tile exists", () => {
        renderComponent();

        expect(
          callCanDrop({
            isOver: () => false,
          }),
        ).toBe(false);
      });

      test("returns false if monitor is over and a tile exists", () => {
        renderComponent({ boardSquare: boardSquareFixture() });

        expect(
          callCanDrop({
            isOver: () => true,
          }),
        ).toBe(false);
      });

      test("returns false if monitor is not over and a tile exists", () => {
        renderComponent({ boardSquare: boardSquareFixture() });

        expect(
          callCanDrop({
            isOver: () => false,
          }),
        ).toBe(false);
      });
    });

    describe("drop", () => {
      const callDrop = (boardLocation: BoardLocation) =>
        useDrop.mock.calls[0][0].drop({ id: "id", boardLocation });

      beforeEach(() => {
        renderComponent();
      });

      test("moves tile to new location on board if tile is already on board", () => {
        const boardLocation = { x: 1, y: 1 };
        callDrop(boardLocation);

        expect(mockHandleMoveTileOnBoard).toHaveBeenCalledWith(boardLocation, {
          x: 0,
          y: 0,
        });
      });

      test("moves tile from hand to board if tile is not on board", () => {
        callDrop(null);

        expect(mockHandleMoveTileFromHandToBoard).toHaveBeenCalledWith("id", {
          x: 0,
          y: 0,
        });
      });
    });

    describe("collect", () => {
      const mockIsOver = jest.fn().mockReturnValue("isOver");
      const mockCanDrop = jest.fn().mockReturnValue("canDrop");

      const callUseDrop = () =>
        useDrop.mock.calls[0][0].collect({
          isOver: mockIsOver,
          canDrop: mockCanDrop,
        });

      beforeEach(() => {
        renderComponent();
      });

      test("calls monitor isOver", () => {
        callUseDrop();

        expect(mockIsOver).toHaveBeenCalledWith({ shallow: true });
      });

      test("calls monitor canDrop", () => {
        callUseDrop();

        expect(mockCanDrop).toHaveBeenCalledWith();
      });

      test("returns the expected valud", () => {
        expect(callUseDrop()).toEqual({
          isOver: "isOver",
          canDrop: "canDrop",
        });
      });
    });
  });
});
