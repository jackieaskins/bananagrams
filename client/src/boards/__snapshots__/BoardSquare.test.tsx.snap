// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<BoardSquare /> renders black tile when not validated in any direction 1`] = `
<Styled(MuiBox)
  height="25px"
  width="25px"
>
  <Tile
    boardPosition={
      Object {
        "col": 0,
        "row": 0,
      }
    }
    color="black"
    tile={
      Object {
        "id": "A1",
        "letter": "A",
      }
    }
  />
</Styled(MuiBox)>
`;

exports[`<BoardSquare /> renders green tile when valid in every direction 1`] = `
<Styled(MuiBox)
  height="25px"
  width="25px"
>
  <Tile
    boardPosition={
      Object {
        "col": 0,
        "row": 0,
      }
    }
    color="green"
    tile={
      Object {
        "id": "A1",
        "letter": "A",
      }
    }
  />
</Styled(MuiBox)>
`;

exports[`<BoardSquare /> renders properly when is over and can drop 1`] = `
<Styled(MuiBox)
  className="validDrop"
  height="25px"
  width="25px"
/>
`;

exports[`<BoardSquare /> renders properly with null boardSquare 1`] = `
<Styled(MuiBox)
  height="25px"
  width="25px"
/>
`;

exports[`<BoardSquare /> renders red tile when invalid in any direction 1`] = `
<Styled(MuiBox)
  height="25px"
  width="25px"
>
  <Tile
    boardPosition={
      Object {
        "col": 0,
        "row": 0,
      }
    }
    color="red"
    tile={
      Object {
        "id": "A1",
        "letter": "A",
      }
    }
  />
</Styled(MuiBox)>
`;
